File: internal/controller/rageval/rageval.go
// Package rageval RAG 质量评估 HTTP 控制器。
package rageval

import (
	"context"

	v1 "Fo-Sentinel-Agent/api/rageval/v1"
	rageval "Fo-Sentinel-Agent/internal/service/rageval"
)

type controllerV1 struct{}

// NewV1 返回 rageval 控制器实例。
func NewV1() *controllerV1 {
	return &controllerV1{}
}

// Dashboard 返回 RAG KPI 汇总。
func (c *controllerV1) Dashboard(ctx context.Context, req *v1.DashboardReq) (res *v1.DashboardRes, err error) {
	m, err := rageval.GetDashboard(ctx, req.Window)
	if err != nil {
		return nil, err
	}
	return &v1.DashboardRes{
		SuccessRate:       m.SuccessRate,
		AvgLatencyMs:      m.AvgLatencyMs,
		P95LatencyMs:      m.P95LatencyMs,
		TotalRuns:         m.TotalRuns,
		AvgRetrievedDocs:  m.AvgRetrievedDocs,
		AvgTopScore:       m.AvgTopScore,
		SuccessRateStatus: m.SuccessRateStatus,
		LatencyStatus:     m.LatencyStatus,
		Trends:            m.Trends,
	}, nil
}

// Traces 返回最近 RAG 链路列表。
func (c *controllerV1) Traces(ctx context.Context, req *v1.TracesReq) (res *v1.TracesRes, err error) {
	items, total, err := rageval.ListTraces(ctx, req.Page, req.PageSize, req.Status)
	if err != nil {
		return nil, err
	}
	return &v1.TracesRes{List: items, Total: total}, nil
}

// TraceDetail 返回 Trace 链路详情。
func (c *controllerV1) TraceDetail(ctx context.Context, req *v1.TraceDetailReq) (res *v1.TraceDetailRes, err error) {
	detail, err := rageval.GetTraceDetail(ctx, req.TraceID)
	if err != nil {
		return nil, err
	}
	return &v1.TraceDetailRes{
		TraceID:           detail.TraceID,
		TraceName:         detail.TraceName,
		SessionID:         detail.SessionID,
		QueryText:         detail.QueryText,
		Status:            detail.Status,
		DurationMs:        detail.DurationMs,
		TotalInputTokens:  detail.TotalInputTokens,
		TotalOutputTokens: detail.TotalOutputTokens,
		EstimatedCostCNY:  detail.EstimatedCostCNY,
		StartTime:         detail.StartTime,
		Nodes:             detail.Nodes,
		FeedbackVote:      detail.FeedbackVote,
	}, nil
}

// DeleteTrace 删除 Trace 链路记录。
func (c *controllerV1) DeleteTrace(ctx context.Context, req *v1.DeleteTraceReq) (res *v1.DeleteTraceRes, err error) {
	return &v1.DeleteTraceRes{}, rageval.DeleteTrace(ctx, req.TraceID)
}

// Feedback 提交消息反馈。
func (c *controllerV1) Feedback(ctx context.Context, req *v1.FeedbackReq) (res *v1.FeedbackRes, err error) {
	return &v1.FeedbackRes{}, rageval.SubmitFeedback(ctx, req.SessionID, req.MessageIndex, req.Vote, req.Reason)
}

// FeedbackStats 返回反馈统计。
func (c *controllerV1) FeedbackStats(ctx context.Context, req *v1.FeedbackStatsReq) (res *v1.FeedbackStatsRes, err error) {
	stats, err := rageval.GetFeedbackStats(ctx)
	if err != nil {
		return nil, err
	}
	return &v1.FeedbackStatsRes{
		LikeRate:    stats.LikeRate,
		DislikeRate: stats.DislikeRate,
		NoVoteRate:  stats.NoVoteRate,
		Total:       stats.Total,
		Recent:      stats.Recent,
	}, nil
}

File: web/src/services/rageval.ts
import api from './api'

export interface TrendPoint {
  timestamp: string
  success_rate: number
  avg_latency_ms: number
}

export interface DashboardMetrics {
  success_rate: number
  avg_latency_ms: number
  p95_latency_ms: number
  total_runs: number
  avg_retrieved_docs: number  // P1
  avg_top_score: number       // P1
  success_rate_status: 'good' | 'warning' | 'bad'
  latency_status: 'good' | 'warning' | 'bad'
  trends: TrendPoint[]
}

export interface TraceItem {
  trace_id: string
  trace_name: string
  session_id: string
  status: string
  duration_ms: number
  start_time: string
  feedback_vote: number  // 0=无 1=赞 -1=踩
}

// P0: Trace 节点树
export interface TraceNodeItem {
  node_id: string
  parent_node_id?: string
  depth: number
  node_type: string
  node_name: string
  status: string
  duration_ms: number
  error_message?: string
  model_name?: string
  input_tokens?: number
  output_tokens?: number
  cost_usd?: number
  cache_hit?: boolean
  final_top_k?: number
  doc_count?: number
  avg_vector_score?: number
  max_vector_score?: number
  rerank_used?: boolean
  avg_rerank_score?: number
  retrieved_docs?: string
  children?: TraceNodeItem[]
}

// P0: Trace 详情
export interface TraceDetail {
  trace_id: string
  trace_name: string
  session_id: string
  query_text: string
  status: string
  duration_ms: number
  total_input_tokens: number
  total_output_tokens: number
  estimated_cost_usd: number
  start_time: string
  nodes: TraceNodeItem[]
  feedback_vote: number
}

export interface RecentFeedback {
  vote: number
  reason?: string
  created_at: string
}

export interface FeedbackStats {
  like_rate: number
  dislike_rate: number
  no_vote_rate: number
  total: number
  recent: RecentFeedback[]
}

interface ApiWrap<T> { data: T }

export const ragevalService = {
  async getDashboard(window = '24h'): Promise<DashboardMetrics> {
    const res = await api.get<ApiWrap<DashboardMetrics>>('/rageval/v1/dashboard', { params: { window } })
    return res.data.data ?? {
      success_rate: 0, avg_latency_ms: 0, p95_latency_ms: 0, total_runs: 0,
      avg_retrieved_docs: 0, avg_top_score: 0,
      success_rate_status: 'good', latency_status: 'good', trends: [],
    }
  },

  async listTraces(params?: {
    page?: number
    pageSize?: number
    status?: string
  }): Promise<{ list: TraceItem[]; total: number }> {
    const res = await api.get<ApiWrap<{ list: TraceItem[]; total: number }>>('/rageval/v1/traces', {
      params: {
        page: params?.page ?? 1,
        page_size: params?.pageSize ?? 5,
        status: params?.status ?? '',
      },
    })
    return { list: res.data.data?.list ?? [], total: res.data.data?.total ?? 0 }
  },

  async getTraceDetail(traceId: string): Promise<TraceDetail> {
    const res = await api.get<ApiWrap<TraceDetail>>('/rageval/v1/traces/detail', { params: { trace_id: traceId } })
    return res.data.data!
  },

  async deleteTrace(traceId: string): Promise<void> {
    await api.delete('/rageval/v1/traces', { data: { trace_id: traceId } })
  },

  async submitFeedback(sessionId: string, messageIndex: number, vote: 1 | -1 | 0, reason?: string): Promise<void> {
    await api.post('/rageval/v1/feedback', {
      session_id: sessionId,
      message_index: messageIndex,
      vote,
      reason: reason ?? '',
    })
  },

  async getFeedbackStats(): Promise<FeedbackStats> {
    const res = await api.get<ApiWrap<FeedbackStats>>('/rageval/v1/feedback_stats')
    return res.data.data!
  },
}
